### DISCUSSION TITLE
How to change C++ defalut compiler?

### DISCUSSION THREAD

**Discussion by saeidex** [zed-industries/zed#8299] [Feature Requests]
I want to change the defalut clang compiler to g++(gcc). How can I achive this?

![Screenshot 2024-02-24 at 1 14 16 AM](https://github.com/zed-industries/zed/assets/112776380/19281421-57ce-42da-874e-7fecf6ade33b)

**Comment by Moshyfawn** [comment]
Check out #4295 and be sure to +👍 if it's something you'd like to see.

**Reply by renjuashokan** [reply]
Sorry, I didnt understand this.
I want to set `g++ -std=c++17` as the c++ compiler, how to do this?

**Comment by NikChouhan** [comment]
Why is this question unanswered since Feb? Are there no solutions at all? It's a dealbreaker if that's the case ngl.

**Comment by vijaydharmaji29** [comment]
# Zed C++ Compiler Configuration

### Follow This

1. Open command palette (Cmd + Shift + P / Ctrl + Shift + P) on Zed
2. Open settings by typing `zed: Open Settings`
3. In your `settings.json` file add or update to include the following

```json
"lsp": {
    "clangd": {
      "binary": {
        "path": "/path/to/compiler",
        "arguments": [
          "--background-index",
          "--compile-commands-dir=build",
          "--clang-tidy",
          "--clang-tidy-checks=*",
          "--header-insertion=never",
          "--header-insertion-decorators=none"
        ]
      }
    }
  }
```

To check the location of your compiler you can run the following in your terminal depending on which compiler’s location you want to get

`which g++`

`which gcc`

`which clang`

**Reply by fus38xid** [reply]
Thank you very much. This works as of Oct 7 :)

**Reply by ugreg** [reply]
@vijaydharmaji29 @fus38xid I am trying to setup my c++ default compiler to C++ 11, any guidance. Right now I am facing this error:

```
Language server error: clangd

oneshot canceled
-- stderr--
clang++: error: unsupported option '--background-index'
clang++: error: unsupported option '--compile-commands-dir=build'
clang++: error: unsupported option '--clang-tidy'
clang++: error: unsupported option '--clang-tidy-checks=*'
clang++: error: unsupported option '--header-insertion=never'
clang++: error: unsupported option '--header-insertion-decorators=none'
clang++: error: no input files
```

I haev `g++` here:
![imagen](https://github.com/user-attachments/assets/f7e588ca-4572-49f0-b23e-555f5e4ba75e)

My [settings.json](https://github.com/user-attachments/files/18745628/settings.json).

**Reply by joeltheai** [reply]
i have the same issue now

**Reply by katara-Jayprakash** [reply]
> @vijaydharmaji29 @fus38xid I am trying to setup my c++ default compiler to C++ 11, any guidance. Right now I am facing this error:
> 
> ```
> Language server error: clangd
> 
> oneshot canceled
> -- stderr--
> clang++: error: unsupported option '--background-index'
> clang++: error: unsupported option '--compile-commands-dir=build'
> clang++: error: unsupported option '--clang-tidy'
> clang++: error: unsupported option '--clang-tidy-checks=*'
> clang++: error: unsupported option '--header-insertion=never'
> clang++: error: unsupported option '--header-insertion-decorators=none'
> clang++: error: no input files
> ```
> 
> I haev `g++` here: ![imagen](https://private-user-images.githubusercontent.com/8151045/411842182-f7e588ca-4572-49f0-b23e-555f5e4ba75e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjIyNDIwOTAsIm5iZiI6MTc2MjI0MTc5MCwicGF0aCI6Ii84MTUxMDQ1LzQxMTg0MjE4Mi1mN2U1ODhjYS00NTcyLTQ5ZjAtYjIzZS01NTVmNWU0YmE3NWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTEwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMDRUMDczNjMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjMzNGFhYmI2NzIzYjUwZDNmMjg1YzVjMGJiMGQzNjIwOGQzZWU3MWU5MTI2ZjlhNWQ2OWJlMTUxMzdmOGQ0NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ytJn-nHpv5mR4kspf-x5FHfMxKGYiLHv2onZrOk6GvE)
> 
> My [settings.json](https://github.com/user-attachments/files/18745628/settings.json).


    "lsp": {
      "g++": {
        "binary": {
          "path": "/opt/homebrew/bin/g++-15",
          "arguments": [
            "-std=c++17",
            "-O2",
            "-Wall"
          ]
        }
      }
    }
    
this should be the solution for this i am using mac  btw so change it according the path according to it;

**Comment by notorand-it** [comment]
So, let's say I need to use `riscv32-unknown-elf-gcc` for syntax checking GCC dialect and RISC-V inline  assembly with Zed.
What would it be like the `lsp.clangd` object?

**Comment by homocodian** [comment]
If you're still wondering how to set up Zed for C++ on Windows using MSYS2 GCC, add the following configuration to your settings.json file.
```json
"lsp": {
    "clangd": {
      "binary": {
        "path": "C:\\msys64\\ucrt64\\bin\\clangd.exe",  // replace the paths with your own
        "arguments": [
          "--query-driver=C:\\msys64\\ucrt64\\bin\\gcc.exe,C:\\msys64\\ucrt64\\bin\\g++.exe",  // replace the paths with your own
          "--background-index",
          "--clang-tidy",
          "--header-insertion=never",
          "-j=4",
        ],
      },
    },
  },
```
